src/v_video.c

```diff
@@ -849,9 +871,39 @@ void V_ScreenShot(char *format)
 #ifdef HAVE_LIBPNG
     if (png_screenshots)
     {
-    WritePNGfile(lbmname, I_VideoBuffer,
-                 SCREENWIDTH, SCREENHEIGHT,
-                 W_CacheLumpName (DEH_String("PLAYPAL"), PU_CACHE));
+        byte * buf, * oldbuf;
+        int oldpitch;
+
+        // Prepare a temporary buffer to write the screenshot to
+        buf = malloc(screenshot_width * screenshot_height);
+
+        if(!buf)
+        {
+            return;
+        }
+
+        // Ideally we'd do this only once, but V_Init is called too early
+        // and screenshot_width/height haven't been read from the config
+        V_InitScreenShotScaler();
+
+        // back up the scaling system's current output buffer settings
+        oldbuf = I_GetScaleDestBuf();
```

Comment:
What I meant was: let's refactor so that nothing calls ->DrawScreen() except code in i_scale.c. Then we have an API function (I called it a "wrapper" around DrawScreen) that sets buf and pitch and then calls it.